source/slang/slang-check-decl.cpp

```diff
@@ -9308,8 +9308,272 @@ namespace Slang
+        ctor->parentDecl = decl;
+        ctor->loc = decl->loc;
+        ctor->closingSourceLoc = ctor->loc;
+        ctor->nameAndLoc.name = ctorName;
+        ctor->nameAndLoc.loc = ctor->loc;
+        ctor->returnType.type = visitor->calcThisType(makeDeclRef(decl));
+        auto body = m_astBuilder->create<BlockStmt>();
+        body->scopeDecl = m_astBuilder->create<ScopeDecl>();
+        body->scopeDecl->ownedScope = m_astBuilder->create<Scope>();
+        body->scopeDecl->ownedScope->parent = visitor->getScope(ctor);
+        body->scopeDecl->parentDecl = ctor;
+        body->scopeDecl->loc = ctor->loc;
+        body->scopeDecl->closingSourceLoc = ctor->loc;
+        body->closingSourceLoc = ctor->closingSourceLoc;
+        ctor->body = body;
+        body->body = m_astBuilder->create<SeqStmt>();
+        decl->addMember(ctor);
+        return ctor;
+    }
+
+    static SeqStmt* legalizeCtorBlockToASeqStmt(ASTBuilder* m_astBuilder, BlockStmt* stmt)
+    {
+        // It is possible BlockStmt has a child with the type of
+        // `ExpressionStmt` if an existing constructor has only 1
+        // expression. This would be a senario we need to
+        // put the `ExpressionStmt` inside a `SeqStmt`.
+        // We need to ensure we are adding to a constructor which can
+        // handle multiple statments since we may be adding field `init`
+        // expressions and function calls to the constructor.
+        if (!as<SeqStmt>(stmt->body))
+        {
+            auto tmpExpr = stmt->body;
+            auto seqStmt = m_astBuilder->create<SeqStmt>();
+            seqStmt->stmts.add(tmpExpr);
+            stmt->body = seqStmt;
+            return seqStmt;
+        }
+        return as<SeqStmt>(stmt->body);
+    }
+    static void legalizeCtorOverloadRank(ASTBuilder* m_astBuilder, ConstructorDecl* ctor, Facet::DirectnessVal rank)
+    {
+        // It is possible a "this" has a "super". Since we are creating a new constructor
+        // function, if there is a "super", there may be a compile error because a derived
+        // struct can use a "super"s constructor.
+        // To solve this issue and produce C++ like behavior we need to add a overload
+        // rank if not present already to prioritize the default constructor of a derived
+        // struct.
+        if (auto overloadRank = ctor->findModifier<OverloadRankAttribute>())
+        {
+            overloadRank->rank = (int32_t(rank) < overloadRank->rank) ? int32_t(rank) : overloadRank->rank;
+            return;
+        }
+
+        auto overloadRankMod = m_astBuilder->create<OverloadRankAttribute>();
+        overloadRankMod->rank = rank;
+        addModifier(ctor, overloadRankMod);
+    }
     void SemanticsDeclAttributesVisitor::visitStructDecl(StructDecl* structDecl)
     {
+        // We need the data of our (maybe) derived "this" and (if exists) "super".
+        // We also need the total inheritance depth to properly set overload ranks
```

Comment:
Why can't we just look at all InheritanceDecl members and call the base types constructor? That way we don't need to go all the way down to every leaf super type.